Missing  SOPN Pages

VirginiaDooley opened the pull request:
Original conversation is here; I had to force push and didn't want to lose the valuable comments. https://github.com/DemocracyClub/yournextrepresentative/pull/1404

co-authored with @symroe @michaeljcollinsuk 

Ref https://github.com/DemocracyClub/yournextrepresentative/issues/882

This work addresses an edge case where the **SOPN has identical size/format headers**.

- [x] Fixes top page confusion when the headings and wards are the same
- [x] Refactors how headers compared against each other 
- [x] Renders fully parsed sopn results in bulk add person form
- [x] Fixes parsing tables when header like rows exist further down the table as a result of appending in the previous step

michaeljcollinsuk commented on ynr/apps/official_documents/views.py:
```diff
@@ -48,6 +48,8 @@ def form_valid(self, form):
         self.object = form.save()
         try:
+            if hasattr(self.object.ballot, "rawpeople"):
+                self.object.ballot.rawpeople.all().delete()
```
The `rawpeople` relation is a OneToOne so called `all()` will raise an error so this should be `self.object.ballot.rawpeople.delete()`

Also I can't quite remember the context of this change, but perhaps we might only want to delete the raw people if the SOPN was parsed successfully? As when it fails the raw people might be useful? If so we could move this to just above the `except` line

symroe commented on ynr/apps/sopn_parsing/helpers/pdf_helpers.py:
```diff
@@ -22,46 +25,79 @@
+                page_numbers.append(page.page_number)
+                page.matched = True
+                previous_page = page
+        return ",".join([str(page_num) for page_num in page_numbers])
```
Thinking about this, it would be good to assert that the pages are consecutive in the list. If they're not, I think we should raise a `ValueError`.

You can check this in a new method on this class, with logic a bit like:
```py
pages = [1,2,3,4,5,13]

lower = min(pages)
upper = max(pages) + 1

assert pages == range(lower, upper)
```

symroe commented on ynr/apps/sopn_parsing/helpers/pdf_helpers.py:
```diff
@@ -69,7 +69,19 @@ def match_ballot_to_pages(self, ballot: Ballot) -> str:
+                raise ValueError("Page numbers are not consecutive")
+
+    @property
+    def contains_consecutive_pages(self):
```
This needs to check the matched pages are consecutive (e.g page_numbers). The document pages are always consecutive by definition, as we assign page numbers in order.

contains_consecutive_pages should accept a list as an arg and check that list, but the logic above can stay more or less as is


(left a previous comment on a now missing commit, sorry for the duplicate!)

VirginiaDooley replied:
Of course 🤦‍♀️ 

I thought pushing with `--force-with-lease` would save the comments :/

michaeljcollinsuk commented on ynr/apps/sopn_parsing/helpers/parse_tables.py:
```diff
@@ -264,6 +272,12 @@ def parse_raw_data_for_ballot(ballot):
                     data = data.drop(data.index[i])
                 except IndexError:
                     break
+        else:
```
Need a refresher on what this is fixing - might it be not necessary since Sym's changes?

michaeljcollinsuk replied:
Ah ok I _think_ I understand

michaeljcollinsuk commented on ynr/apps/sopn_parsing/helpers/pdf_helpers.py:
```diff
@@ -19,20 +21,106 @@
+        if len(self.pages) == 1 and len(self.unmatched_documents) == 1:
+            return [(self.unmatched_documents.pop(), "all")]
+
+        docs_by_sorted_ballot_label = sorted(
```
Not sure this is needed as ordering is already applied when we get the `OfficialDocument` objects earlier on https://github.com/DemocracyClub/yournextrepresentative/pull/1659/files#

symroe replied:
Yep, I just wanted to make really sure they were sorted. It feels better to do this "inside the black box" rather than trusting it to be sorted in advance.

In fact, I wonder if taking a file and a source would be better for this class, and then we can do the query in-class? What do you think?

michaeljcollinsuk replied:
Yeah I think that's a good idea - that would allow us to simplify the logic  in `extract_pages_for_single_document` also

symroe commented on ynr/apps/sopn_parsing/helpers/pdf_helpers.py:
```diff
@@ -102,27 +170,82 @@ def get_page_heading(self):
+            == previous_page_heading_up_to_ward_name
+        )
+
+        # TO DO: multipage sopns with multiple wards where headings
```
Isn't what you describe the correct behaviour in this case? The pages are for different ballots, so `is_continuation_page` shouldn't return True (we might want it to always return False rather than None, but that's not the issue here).

michaeljcollinsuk replied:
Yes I agree - went through this with Virginia and cleared up the issues she was having (because it was a by-election SOPN all pages were being returned incorrectly)

VirginiaDooley commented:
Really nice improvements, so much easier to read. 🥇

michaeljcollinsuk commented:
I rebased with master now that the the SOPN tool PR was merged. Running the SOPN test caused on error [due to changes in this commit](https://github.com/DemocracyClub/yournextrepresentative/pull/1659/commits/113ad79b01246f00ff4921bfc703f7e00a6f93aa). So for now I reverted that commit, so that I could run `make test-sopns` which results in 247 RawPeople the same as on master.

@VirginiaDooley It was so long ago that I've forgot what that commit is trying to do but feel free to revisit it if you want

VirginiaDooley commented:
> I rebased with master now that the the SOPN tool PR was merged. Running the SOPN test caused on error [due to changes in this commit](https://github.com/DemocracyClub/yournextrepresentative/pull/1659/commits/983842599e552afa865489deba64e4f92f32bbc1). So for now I reverted that commit, so that I could run `make test-sopns` which results in 247 RawPeople the same as on master.
> 
> @VirginiaDooley It was so long ago that I've forgot what that commit is trying to do but feel free to revisit it if you want

I didn't run into any error with that commit, but I did get an error related to "Page numbers are not consecutive" in `pdf_helpers.py`
